### PR TITLE
Add safe margin parameter to avoid rebuffering interruptions when clearing the buffer

### PR DESCRIPTION
**Pull request by wasp898**
The goal of this PR is to introduce an optional parameter that allows the buffer to be cleared but not entirely.

The idea is to keep a safe margin that avoids rebuffering events but allows for fast switches. The value of this parameter should be computed and provided by a custom ABR manager. Nothing changes if the value is not provided of course (e.g. SimpleAbrManager).

Since the buffer could be busy, another parameter had to be introduced to save the value of the safe margin when the call to clear the buffer is deferred to a later update.

Something similar was present in Shaka v1, but it was removed with the redesign.

### CODE REVIEW

**Comment by wasp898** [comment]
I signed it!

**Comment by joeyparrish** [comment]
The trouble with this is that clearing media ahead of the playhead causes stuttering issues on some browsers.  That is why we don't stopped doing that in v2.

Have you tested this on multiple browsers?  Which browsers does your application intend to support?

**Comment by wasp898** [comment]
Thanks a lot for your reply.

These few days I've spent some time testing it on every browser for desktop. Issues occur when you clear too close to the playhead. As long as the margin is computed appropriately, I don't see any artifect or stuttering. Playback goes smoothly.

Here's what I tried to do in detail. After computing (approximately) the next segment boundary, I tried clearing from that point on when switching variant. This caused big problems already (because of a lack of accuracy probably), so I tried adding 0.2 seconds to such margin, and it was enough to solve the issue on most browsers except mainly for Chrome. Finally, I tried to increase the margin by adding the duration of a segment, and I stopped getting stuttering altogether. As long as the network state is good though. I mean, the only stutters that I noticed have happened because of buffer depletion, but that's a different issue.

Bottom line is: I think two segments worth of buffer is a safe minimum to consider when computing the margin; and then the final value should be computed according to the bandwidth estimate I guess, to avoid depletion.

I have found the commit where the support was dropped: https://github.com/google/shaka-player/commit/385a501d0dd2aa31f55d5425441baca86ac4d982


p.s. I've tested with Chrome, Firefox (52 and 57), Safari, Opera, Edge, IE 11. I also tried different segment sizes: 1, 4 and 6 seconds. Tested under good network conditions, mainly. Some tests under bandwidth cap as well.

**Comment by joeyparrish** [comment]
Thank you for doing such detailed testing.  At this point, our plates are pretty full for the rest of the year to finish v2.3.0.  We will look into this PR in January when we start work on v2.4.0.  Please accept our apologies for the delay, and thank you for your contribution!

**Comment by wasp898** [comment]
Thank you!

File: lib/media/streaming_engine.js [hunk 1]
```diff
@@ -991,7 +1002,8 @@ shaka.media.StreamingEngine.prototype.onUpdate_ = function(mediaState) {
   if (mediaState.waitingToClearBuffer) {
     // Note: clearBuffer_() will schedule the next update.
     shaka.log.debug(logPrefix, 'skipping update and clearing the buffer');
-    this.clearBuffer_(mediaState, mediaState.waitingToFlushBuffer);
+    this.clearBuffer_(mediaState, mediaState.waitingToFlushBuffer,
+        mediaState.deferredClearBufferOffset);
```
Comment (TheModMaker) [inline lib/media/streaming_engine.js]: The style is to keep all the arguments aligned.  So please do either:

```js
this.clearBuffer_(mediaState, mediaState.waitingToFlushBuffer,
                  mediaState.deferredClearBufferOffset);
```
Or
```js
this.clearBuffer_(
    mediaState, mediaState.waitingToFlushBuffer,
    mediaState.deferredClearBufferOffset);
```

**Comment by wasp898** [comment]
Okay, I rebased and fixed the style related issues as requested. No conflicts, and build went fine.

File: lib/media/streaming_engine.js [hunk 2]
```diff
@@ -2160,12 +2177,21 @@ shaka.media.StreamingEngine.prototype.clearBuffer_ =
...
+    let playheadTime = this.playerInterface_.playhead.getTime();
+    let duration = this.playerInterface_.mediaSourceEngine.getDuration();
+    p = this.playerInterface_.mediaSourceEngine.remove(mediaState.type,
+        playheadTime + safeMargin, duration);
+  } else {
+    p = this.playerInterface_.mediaSourceEngine.clear(mediaState.type);
+  }
   p.then(function() {
-    if (!this.destroyed_ && flush) {
+    if (!this.destroyed_ && flush && safeMargin == 0) {
```
Comment (TheModMaker) [inline lib/media/streaming_engine.js]: Since this is dependent on `safeMargin == 0`, you could put this as part of line 2191.

File: lib/media/streaming_engine.js [hunk 3]
```diff
@@ -237,6 +237,7 @@ shaka.media.StreamingEngine.PlayerInterface;
  *   updateTimer: ?number,
  *   waitingToClearBuffer: boolean,
  *   waitingToFlushBuffer: boolean,
+ *   deferredClearBufferOffset: number,
```
Comment (TheModMaker) [inline lib/media/streaming_engine.js]: I think a better name would be `clearBufferSafeMargin`, or something including safeMargin, to be consistent.

File: lib/player.js [hunk 4]
```diff
@@ -133,6 +133,9 @@ shaka.Player = function(video, dependencyInjector) {
   /** @private {boolean} */
   this.deferredVariantClearBuffer_ = false;
 
+  /** @private {number} */
+  this.deferredVariantClearBufferOffset_ = 0;
```
Comment (TheModMaker) [inline lib/player.js]: "Deferred" is fine here, but again, I think it should include safeMargin instead of offset.

File: lib/player.js [hunk 5]
```diff
@@ -1640,9 +1643,11 @@ shaka.Player.prototype.usingEmbeddedTextTrack = function() {
  *
  * @param {shaka.extern.Track} track
  * @param {boolean=} clearBuffer
+ * @param {number=} safeMargin
```
Comment (TheModMaker) [inline lib/player.js]: This deserves a comment for what this will do.  You should also include a part about browser compatibility for having this value be too small.

File: lib/player.js [hunk 6]
```diff
@@ -2961,9 +2970,11 @@ shaka.Player.prototype.onSegmentAppended_ = function() {
  *
  * @param {shaka.extern.Variant} variant
  * @param {boolean=} clearBuffer
+ * @param {number=} safeMargin
  * @private
  */
-shaka.Player.prototype.switch_ = function(variant, clearBuffer = false) {
+shaka.Player.prototype.switch_ = function(
+    variant, clearBuffer = false, safeMargin = 0) {
```
Comment (TheModMaker) [inline lib/player.js]: This is the callback from AbrManager.  If you want to change this, you'll need to change the definitions in `externs/shaka/abr_manager.js` and in SimpleAbrManager to match the new function signature.

File: CONTRIBUTORS [hunk 7]
```diff
@@ -33,6 +33,7 @@ Donato Borrello <donato@jwplayer.com>
 Duc Pham <duc.pham@edgeware.tv>
 Esteban Dosztal <edosztal@gmail.com>
 François Beaufort <fbeaufort@google.com>
+Giuseppe Samela <giuseppe.samela@gmail.com>
```
Comment (TheModMaker) [inline CONTRIBUTORS]: You should also add an entry to AUTHORS.

File: lib/media/streaming_engine.js (hunk 2 above)
Comment (wasp898) [reply lib/media/streaming_engine.js]: What did you have in mind here? Not sure how to move this without duplicating things after.
```
p.then(function() {
    if (!this.destroyed_ && flush && safeMargin == 0) {
      return this.playerInterface_.mediaSourceEngine.flush(mediaState.type);
    }
  }.bind(this)).then(function() {
    if (this.destroyed_) return;
    shaka.log.debug(logPrefix, 'cleared buffer');
    mediaState.lastStream = null;
    mediaState.lastSegmentReference = null;
    mediaState.clearingBuffer = false;
    mediaState.endOfStream = false;
    this.scheduleUpdate_(mediaState, 0);
  }.bind(this));
```

Comment (TheModMaker) [reply lib/media/streaming_engine.js]:
```js
if (safeMargin) {
  p = ...
} else {
  p = this.playerInterface_.mediaSourceEngine.clear(mediaState.type).then(() => {
    if (!this.destroyed_ && flush)
      return this.playerInterface_.mediaSourceEngine.flush(mediaState.type);
  });
}
p.then(() => {
  if (this.destroyed_) return;
  shaka.log.debug(logPrefix, 'cleared buffer');
  ...
});
```

Comment (wasp898) [reply lib/media/streaming_engine.js]: Oh okay, got it, thanks!

File: lib/player.js (hunk 6 above)
Comment (wasp898) [reply lib/player.js]: In SimpleAbrManager I see that the call to the switch function only passes the variant, without the optional parameters. I kept it this way. Is that okay?
I pushed the other modifications, including the updated abr manager extern.

Comment (TheModMaker) [reply lib/player.js]: That's fine.
